lib/vmpooler/prometheus.rb

```diff
@@ -0,0 +1,240 @@
+          docstring: 'vmpooler machines running',
+          prom_metric_prefix: 'vmpooler_running',
+          param_labels: %i[poolname]
+        },
+        time_to_ready_state: {
+          mtype: M_HISTOGRAM,
+          docstring: 'vmpooler machine time to ready',
+          prom_metric_prefix: 'vmpooler_time_to_ready_state',
+          param_labels: %i[poolname]
+        },
+        clone: {
+          mtype: M_HISTOGRAM,
+          docstring: 'vmpooler machine clone time',
+          prom_metric_prefix: 'vmpooler_clone',
+          param_labels: %i[poolname]
+        },
+        destroy: {
+          mtype: M_HISTOGRAM,
+          docstring: 'vmpooler machine destroy time',
+          prom_metric_prefix: 'vmpooler_destroy',
+          param_labels: %i[poolname]
+        },
+        connection_available: {
+          mtype: M_GAUGE,
+          docstring: 'connection available count',
+          prom_metric_prefix: 'vmpooler_connection_available',
+          param_labels: %i[provider]
+        },
+        connection_waited: {
+          mtype: M_HISTOGRAM,
+          docstring: 'connection wait time',
+          prom_metric_prefix: 'vmpooler_connection_waited',
+          param_labels: %i[provider]
+        }
+      }
+    end
+
+    def initialize(params = {})
+      # Initialiser simply creates the instance - we deliberately don't wish to
+      # touch the registry until Sinatra::Base has been initialised otherwise
+      # confusion reigns with the middleware.
+      @prefix = params['prefix']
+      @logger = params.delete('logger') || Logger.new(STDOUT)
+    end
+
+    def setup_prometheus_metrics
+      @prometheus = Prometheus::Client.registry
+
+      @p_metrics = vmpooler_metrics_table
+      @p_metrics.each do |name, metric_spec|
+        case metric_spec[:mtype]
+        when M_COUNTER
+          metric_class = Prometheus::Client::Counter
+        when M_GAUGE
+          metric_class = Prometheus::Client::Gauge
+        when M_SUMMARY
+          metric_class = Prometheus::Client::Summary
+        when M_HISTOGRAM
+          metric_class = Prometheus::Client::Histogram
+        else
+          raise("Unable to register metric #{name} with metric type #{metric_spec[:mtype]}")
+        end
+
+        if metric_spec.key? :metric_suffixes
+          # Iterate thru the suffixes if provided to register multiple counters here.
+          metric_spec[:metric_suffixes].each do |counter_suffix|
+            metric_name = "#{metric_spec[:prom_metric_prefix]}_#{counter_suffix}"
+            prom_metric = metric_class.new(
+              metric_name.to_sym,
+              docstring: metric_spec[:docstring],
+              labels: metric_spec[:param_labels] + [:vmpooler_instance],
+              preset_labels: { vmpooler_instance: @prefix }
+            )
+            # this feels like a hack... (just repeating what Mikker said ;) )
+            @prometheus.register(prom_metric) unless @prometheus.exist?(prom_metric.name)
```

Comment:
Is this hack actually needed? AKA, does this app actually try to initialize them multiple times?

Reply:
Fixing and also noticed some pasta here  that I've tidied.

Reply:
Tidied it to remove hack - but wondering if I should check if the metric is already there - if it is, we should raise an error.
hmmmm